File: src/yoinkc/inspectors/selinux.py
"""SELinux/Security inspector: mode, modules, booleans, audit rules, FIPS, PAM. File-based + executor."""

import re
from pathlib import Path
from typing import List, Optional

from ..executor import Executor
from ..schema import SelinuxSection
from .._util import debug as _debug_fn, safe_iterdir as _safe_iterdir, make_warning


def _debug(msg: str) -> None:
    _debug_fn("selinux", msg)


def _policy_type(host_root: Path) -> str:
    """Read SELINUXTYPE from /etc/selinux/config, default to 'targeted'."""
    cfg = host_root / "etc/selinux/config"
    try:
        if cfg.exists():
            for line in cfg.read_text().splitlines():
                line = line.strip()
                if line.startswith("SELINUXTYPE="):
                    return line.split("=", 1)[1].strip()
    except (PermissionError, OSError):
        pass
    return "targeted"


def _discover_custom_modules(host_root: Path, policy_type: str) -> List[str]:
    """Discover custom modules from the priority-400 module store.

    Modules at priority 400 were installed locally via ``semodule -i``.
    This is purely filesystem-based — no need for ``semodule`` command.
    """
    local_store = (
        host_root / "etc/selinux" / policy_type / "active/modules/400"
    )
    _debug(f"custom modules: checking {local_store}")
    try:
        if not local_store.is_dir():
            _debug(f"custom modules: {local_store} not a directory or missing")
            return []
    except (PermissionError, OSError) as e:
        _debug(f"custom modules: {local_store} access error: {e}")
        return []

    local_names = []
    for child in _safe_iterdir(local_store):
        if child.is_dir():
            local_names.append(child.name)

    _debug(f"custom modules: found {len(local_names)} in priority-400 store: {local_names}")
    return sorted(local_names)


_BOOL_RE = re.compile(
    r"^(\S+)\s+\((\w+)\s*,\s*(\w+)\)\s+(.*)"
)


def _parse_semanage_booleans(text: str) -> List[dict]:
    """Parse ``semanage boolean -l`` output.

    Returns all booleans where current state differs from the default,
    each as ``{"name": ..., "current": ..., "default": ..., "description": ...}``.
    """
    results: List[dict] = []
    for line in text.splitlines():
        line = line.strip()
        if not line or line.startswith("SELinux boolean"):
            continue
        m = _BOOL_RE.match(line)
        if not m:
            continue
        name, current, default, desc = m.group(1), m.group(2), m.group(3), m.group(4).strip()
        results.append({
            "name": name,
            "current": current,
            "default": default,
            "non_default": current != default,
            "description": desc,
        })
    return results


def _read_booleans_from_fs(host_root: Path) -> List[dict]:
    """Fallback: read boolean runtime values from /sys/fs/selinux/booleans/.

    Returns only booleans whose current value differs from the pending
    (policy-loaded) value.  Without semanage we cannot get descriptions.
    """
    booldir = host_root / "sys/fs/selinux/booleans"
    _debug(f"boolean fs fallback: checking {booldir}")
    if not booldir.is_dir():
        _debug(f"boolean fs fallback: {booldir} not found")
        return []

    results: List[dict] = []
    for f in _safe_iterdir(booldir):
        if not f.is_file():
            continue
        try:
            parts = f.read_text().strip().split()
        except (PermissionError, OSError):
            continue
        if len(parts) >= 2:
            current = "on" if parts[0] == "1" else "off"
            pending = "on" if parts[1] == "1" else "off"
            if current != pending:
                results.append({
                    "name": f.name,
                    "current": current,
                    "default": pending,
                    "non_default": True,
                    "description": "",
                })
    _debug(f"boolean fs fallback: found {len(results)} non-default booleans")
    return results


def run(
    host_root: Path,
    executor: Optional[Executor],
    warnings: Optional[list] = None,
) -> SelinuxSection:
    section = SelinuxSection()
    host_root = Path(host_root)

    selinux_config = host_root / "etc/selinux/config"
    try:
        if selinux_config.exists():
            for line in selinux_config.read_text().splitlines():
                line = line.strip()
                if line.startswith("SELINUX="):
                    section.mode = line.split("=", 1)[1].strip()
                    break
    except (PermissionError, OSError):
        pass

    ptype = _policy_type(host_root)
    _debug(f"policy type: {ptype}")

    # --- Custom modules from priority-400 store (filesystem only) ---
    section.custom_modules = _discover_custom_modules(host_root, ptype)

    # --- Boolean overrides via chroot semanage, with filesystem fallback ---
    if executor:
        # Try the host's own semanage via chroot (works against host policy)
        _debug("trying: chroot /host semanage boolean -l")
        out = executor(["chroot", str(host_root), "semanage", "boolean", "-l"])
        if out.returncode == 0 and out.stdout.strip():
            _debug(f"semanage boolean -l succeeded ({len(out.stdout.splitlines())} lines)")
            section.boolean_overrides = _parse_semanage_booleans(out.stdout)
        else:
            _debug(f"semanage failed (rc={out.returncode}): {out.stderr.strip()[:200]}")
            # Fallback: try reading /sys/fs/selinux/booleans/ from the host
            fallback = _read_booleans_from_fs(host_root)
            section.boolean_overrides = fallback
            booldir = host_root / "sys/fs/selinux/booleans"
            if not booldir.is_dir() and warnings is not None:
                warnings.append(make_warning(
                    "selinux",
                    "SELinux boolean override detection unavailable — semanage failed and /sys/fs/selinux/booleans not accessible.",
                ))

    # --- Custom fcontext rules -----------------------------------------------
    # Try semanage fcontext -l -C (custom only) via chroot; fall back to
    # reading file_contexts.local from the policy store.
    if executor:
        out = executor(["chroot", str(host_root), "semanage", "fcontext", "-l", "-C"])
        if out.returncode == 0 and out.stdout.strip():
            for line in out.stdout.strip().splitlines():
                line = line.strip()
                if line and not line.startswith("SELinux"):
                    section.fcontext_rules.append(line)
            _debug(f"fcontext: {len(section.fcontext_rules)} custom rules from semanage")
        else:
            _debug("semanage fcontext failed, trying file_contexts.local")
    if not section.fcontext_rules:
        fc_local = host_root / "etc/selinux" / ptype / "contexts/files/file_contexts.local"
        try:
            if fc_local.exists():
                for line in fc_local.read_text().splitlines():
                    line = line.strip()
                    if line and not line.startswith("#"):
                        section.fcontext_rules.append(line)
                _debug(f"fcontext: {len(section.fcontext_rules)} rules from file_contexts.local")
        except (PermissionError, OSError) as e:
            _debug(f"fcontext: cannot read file_contexts.local: {e}")

    audit_d = host_root / "etc/audit/rules.d"
    if audit_d.exists():
        for f in _safe_iterdir(audit_d):
            if f.is_file():
                section.audit_rules.append(str(f.relative_to(host_root)))

    fips_path = host_root / "proc/sys/crypto/fips_enabled"
    try:
        if fips_path.exists():
            section.fips_mode = fips_path.read_text().strip() == "1"
    except (PermissionError, OSError):
        pass

    pam_d = host_root / "etc/pam.d"
    if pam_d.exists():
        for f in _safe_iterdir(pam_d):
            if f.is_file():
                section.pam_configs.append(str(f.relative_to(host_root)))

    return section

File: cmd/inspectah/internal/inspector/selinux.go
// SELinux/Security inspector: mode, custom modules, boolean overrides,
// fcontext rules, port labels, audit rules, FIPS mode, PAM configs.
//
// Combines executor-based commands (getenforce, semanage) with filesystem
// reads for fallback paths and file scanning.
package inspector

import (
	"path/filepath"
	"regexp"
	"sort"
	"strings"

	"github.com/marrusl/inspectah/cmd/inspectah/internal/schema"
)

// ---------------------------------------------------------------------------
// Public API
// ---------------------------------------------------------------------------

// SelinuxOptions configures the SELinux inspector.
type SelinuxOptions struct {
	// RpmOwnedPaths is the set of paths owned by RPM packages. When
	// non-nil, audit rule files and PAM configs owned by RPM are skipped.
	RpmOwnedPaths map[string]bool
}

// RunSelinux runs the SELinux/security inspector and returns the populated
// section, accumulated warnings, and any fatal error.
func RunSelinux(exec Executor, opts SelinuxOptions) (*schema.SelinuxSection, []Warning, error) {
	var warnings []Warning
	section := &schema.SelinuxSection{
		CustomModules:    []string{},
		BooleanOverrides: []map[string]interface{}{},
		FcontextRules:    []string{},
		AuditRules:       []string{},
		PamConfigs:       []string{},
		PortLabels:       []schema.SelinuxPortLabel{},
	}

	collectSELinuxMode(exec, section)
	ptype := readPolicyType(exec)
	collectCustomModules(exec, section, ptype)
	collectBooleanOverrides(exec, section, &warnings)
	collectFcontextRules(exec, section, ptype)
	collectPortLabels(exec, section)
	collectAuditRules(exec, section, opts.RpmOwnedPaths)
	collectFIPSMode(exec, section)
	collectPAMConfigs(exec, section, opts.RpmOwnedPaths)

	return section, warnings, nil
}

// ---------------------------------------------------------------------------
// SELinux mode
// ---------------------------------------------------------------------------

// collectSELinuxMode reads the SELinux mode from /etc/selinux/config.
func collectSELinuxMode(exec Executor, section *schema.SelinuxSection) {
	content, err := exec.ReadFile("/etc/selinux/config")
	if err != nil {
		return
	}
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "SELINUX=") {
			section.Mode = strings.TrimPrefix(line, "SELINUX=")
			return
		}
	}
}

// ---------------------------------------------------------------------------
// Policy type
// ---------------------------------------------------------------------------

// readPolicyType reads SELINUXTYPE from /etc/selinux/config, defaulting
// to "targeted".
func readPolicyType(exec Executor) string {
	content, err := exec.ReadFile("/etc/selinux/config")
	if err != nil {
		return "targeted"
	}
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if strings.HasPrefix(line, "SELINUXTYPE=") {
			val := strings.TrimPrefix(line, "SELINUXTYPE=")
			if val != "" {
				return val
			}
		}
	}
	return "targeted"
}

// ---------------------------------------------------------------------------
// Custom modules (priority-400 store)
// ---------------------------------------------------------------------------

// collectCustomModules discovers custom SELinux modules from the
// priority-400 module store. Modules at priority 400 were installed
// locally via "semodule -i". Purely filesystem-based.
func collectCustomModules(exec Executor, section *schema.SelinuxSection, policyType string) {
	storePath := filepath.Join("/etc/selinux", policyType, "active/modules/400")
	entries, err := exec.ReadDir(storePath)
	if err != nil {
		return
	}

	var names []string
	for _, e := range entries {
		if e.IsDir() {
			names = append(names, e.Name())
		}
	}
	sort.Strings(names)
	section.CustomModules = names
}

// ---------------------------------------------------------------------------
// Boolean overrides
// ---------------------------------------------------------------------------

// boolRE matches semanage boolean -l output lines:
//   name  (current , default)  description
var boolRE = regexp.MustCompile(`^(\S+)\s+\((\w+)\s*,\s*(\w+)\)\s+(.*)`)

// parseSemanageBooleans parses "semanage boolean -l" output and returns
// booleans where current state differs from default.
func parseSemanageBooleans(text string) []map[string]interface{} {
	var results []map[string]interface{}
	for _, line := range strings.Split(text, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "SELinux boolean") {
			continue
		}
		m := boolRE.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		name, current, defaultVal, desc := m[1], m[2], m[3], strings.TrimSpace(m[4])
		if current != defaultVal {
			results = append(results, map[string]interface{}{
				"name":        name,
				"current":     current,
				"default":     defaultVal,
				"non_default": true,
				"description": desc,
			})
		}
	}
	return results
}

// readBoolsFromFS is a fallback that reads boolean runtime values from
// /sys/fs/selinux/booleans/. Returns only booleans whose current value
// differs from the pending (policy-loaded) value.
func readBoolsFromFS(exec Executor) []map[string]interface{} {
	boolDir := "/sys/fs/selinux/booleans"
	entries, err := exec.ReadDir(boolDir)
	if err != nil {
		return nil
	}

	var results []map[string]interface{}
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		content, err := exec.ReadFile(filepath.Join(boolDir, e.Name()))
		if err != nil {
			continue
		}
		parts := strings.Fields(strings.TrimSpace(content))
		if len(parts) < 2 {
			continue
		}
		current := "off"
		if parts[0] == "1" {
			current = "on"
		}
		pending := "off"
		if parts[1] == "1" {
			pending = "on"
		}
		if current != pending {
			results = append(results, map[string]interface{}{
				"name":        e.Name(),
				"current":     current,
				"default":     pending,
				"non_default": true,
				"description": "",
			})
		}
	}
	return results
}

// collectBooleanOverrides tries semanage boolean -l via chroot, then
// falls back to reading /sys/fs/selinux/booleans/.
func collectBooleanOverrides(exec Executor, section *schema.SelinuxSection, warnings *[]Warning) {
	hostRoot := exec.HostRoot()
	res := exec.Run("chroot", hostRoot, "semanage", "boolean", "-l")
	if res.ExitCode == 0 && strings.TrimSpace(res.Stdout) != "" {
		section.BooleanOverrides = parseSemanageBooleans(res.Stdout)
		return
	}

	// Fallback to filesystem
	fallback := readBoolsFromFS(exec)
	if fallback != nil {
		section.BooleanOverrides = fallback
		return
	}

	// Neither method worked — check if the bool dir exists
	if !exec.FileExists("/sys/fs/selinux/booleans") {
		*warnings = append(*warnings, makeWarning(
			"selinux",
			"SELinux boolean override detection unavailable — semanage failed and /sys/fs/selinux/booleans not accessible.",
		))
	}
}

// ---------------------------------------------------------------------------
// Custom fcontext rules
// ---------------------------------------------------------------------------

// collectFcontextRules tries "semanage fcontext -l -C" via chroot, then
// falls back to reading file_contexts.local from the policy store.
func collectFcontextRules(exec Executor, section *schema.SelinuxSection, policyType string) {
	hostRoot := exec.HostRoot()
	res := exec.Run("chroot", hostRoot, "semanage", "fcontext", "-l", "-C")
	if res.ExitCode == 0 && strings.TrimSpace(res.Stdout) != "" {
		for _, line := range strings.Split(strings.TrimSpace(res.Stdout), "\n") {
			line = strings.TrimSpace(line)
			if line != "" && !strings.HasPrefix(line, "SELinux") {
				section.FcontextRules = append(section.FcontextRules, line)
			}
		}
		if len(section.FcontextRules) > 0 {
			return
		}
	}

	// Fallback: read file_contexts.local
	fcLocal := filepath.Join("/etc/selinux", policyType, "contexts/files/file_contexts.local")
	content, err := exec.ReadFile(fcLocal)
	if err != nil {
		return
	}
	for _, line := range strings.Split(content, "\n") {
		line = strings.TrimSpace(line)
		if line != "" && !strings.HasPrefix(line, "#") {
			section.FcontextRules = append(section.FcontextRules, line)
		}
	}
}

// ---------------------------------------------------------------------------
// Custom port labels
// ---------------------------------------------------------------------------

// portRE matches semanage port -l -C output lines:
//   type  protocol  port[,port...]
var portRE = regexp.MustCompile(`(?i)^(\S+)\s+(tcp|udp)\s+([\d,\-\s]+)`)

// parseSemanagePorts parses "semanage port -l -C" output into port labels.
func parseSemanagePorts(text string) []schema.SelinuxPortLabel {
	var results []schema.SelinuxPortLabel
	for _, line := range strings.Split(text, "\n") {
		line = strings.TrimSpace(line)
		if line == "" || strings.HasPrefix(line, "SELinux") {
			continue
		}
		m := portRE.FindStringSubmatch(line)
		if m == nil {
			continue
		}
		portType := m[1]
		protocol := strings.ToLower(m[2])
		portsRaw := strings.TrimSpace(m[3])

		// Emit one entry per port/range in comma-separated list
		for _, port := range strings.Split(portsRaw, ",") {
			port = strings.TrimSpace(port)
			if port != "" {
				results = append(results, schema.SelinuxPortLabel{
					Protocol: protocol,
					Port:     port,
					Type:     portType,
					Include:  true,
				})
			}
		}
	}
	return results
}

// collectPortLabels runs semanage port -l -C via chroot.
func collectPortLabels(exec Executor, section *schema.SelinuxSection) {
	hostRoot := exec.HostRoot()
	res := exec.Run("chroot", hostRoot, "semanage", "port", "-l", "-C")
	if res.ExitCode == 0 && strings.TrimSpace(res.Stdout) != "" {
		section.PortLabels = parseSemanagePorts(res.Stdout)
	}
}

// ---------------------------------------------------------------------------
// Audit rules
// ---------------------------------------------------------------------------

// collectAuditRules scans /etc/audit/rules.d/ for custom audit rule files,
// skipping files owned by RPM.
func collectAuditRules(exec Executor, section *schema.SelinuxSection, rpmOwned map[string]bool) {
	auditDir := "/etc/audit/rules.d"
	entries, err := exec.ReadDir(auditDir)
	if err != nil {
		return
	}

	var rules []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		rel := filepath.Join("etc/audit/rules.d", e.Name())
		absPath := "/" + rel
		if rpmOwned != nil && rpmOwned[absPath] {
			continue
		}
		rules = append(rules, rel)
	}
	sort.Strings(rules)
	section.AuditRules = rules
}

// ---------------------------------------------------------------------------
// FIPS mode
// ---------------------------------------------------------------------------

// collectFIPSMode reads /proc/sys/crypto/fips_enabled.
func collectFIPSMode(exec Executor, section *schema.SelinuxSection) {
	content, err := exec.ReadFile("/proc/sys/crypto/fips_enabled")
	if err != nil {
		return
	}
	section.FipsMode = strings.TrimSpace(content) == "1"
}

// ---------------------------------------------------------------------------
// PAM configs
// ---------------------------------------------------------------------------

// collectPAMConfigs scans /etc/pam.d/ for custom PAM configuration files,
// skipping files owned by RPM and system-generated exclusions.
func collectPAMConfigs(exec Executor, section *schema.SelinuxSection, rpmOwned map[string]bool) {
	pamDir := "/etc/pam.d"
	entries, err := exec.ReadDir(pamDir)
	if err != nil {
		return
	}

	var configs []string
	for _, e := range entries {
		if e.IsDir() {
			continue
		}
		rel := filepath.Join("etc/pam.d", e.Name())
		absPath := "/" + rel
		if rpmOwned != nil && rpmOwned[absPath] {
			continue
		}
		if isExcludedUnowned(absPath) {
			continue
		}
		configs = append(configs, rel)
	}
	sort.Strings(configs)
	section.PamConfigs = configs
}
